### PR TITLE
When splitting channels, allow coloring and filtering of channels

### PR DESCRIPTION
**Pull request by win-chen**
Our project would like to allow the user to filter out channels from the audio for playing/viewing. The filtering for playing will be done via setFilters. This pr is to add the rest via a new parameter **splitChannelsOptions** and a new method **setFilteredChannels**.

## Short description of changes:
1) Add splitChannelsOptions parameter that applies if splitChannels is true.
Contains the following options:
overlay - draws channels on over each other when splitting
channelColors - defines colors for channels individually
filteredChannels - channel indices to filter (do not draw on the waveform)

2) Add setFilteredChannels method to update the filteredChannels param.

3) Update split channels example to include info on the above.

### Breaking in the external API:
none

### Breaking changes in the internal API:
none

### Todos/Notes:
The splitChannelsOptions currently only works with drawWave.

### Related Issues and other PRs:
coloring:
https://github.com/katspaugh/wavesurfer.js/pull/1889

### CODE REVIEW

**Comment by thijstriemstra** [comment]
this looks similar to #1889, just a different implementation, right @win-chen?

**Comment by win-chen** [comment]
Yea it's similar. The differences are:

#1889 creates two canvases and waveforms in order to color the different waves. This pr adds changes into existing methods to apply colors. 

The shape of the parameters are different. That one adds a waveColor2 and progressColor2 so it allows coloring two channels. This one sets colors using a map from channel index to a { waveColor, progressColor } object so it should work with audio that has more channels than stereo.

This one includes hiding different channels and stacking the wave forms on top of each other which are not part of #1889.

File: src/drawer.multicanvas.js [hunk 1]
```diff
@@ -390,11 +390,13 @@ export default class MultiCanvas extends Drawer {
      * @param {number} start The x-offset of the beginning of the area that
      * should be rendered
      * @param {number} end The x-offset of the end of the area that
+     * @param {channelIndex} channelIndex The channel index of the line drawn
      * should be rendered
      */
```
Comment (thijstriemstra) [inline src/drawer.multicanvas.js:395]: the "should be rendered" belongs to doc of `end` parameter

File: src/drawer.multicanvas.js [hunk 2]
```diff
@@ -457,25 +470,36 @@ export default class MultiCanvas extends Drawer {
...
+                    } 
+
+                    return channels.forEach((channelPeaks, i) => 
+                        this.prepareDraw(channelPeaks, i, start, end, fn, filteredChannels.indexOf(channelPeaks))
+                    );                    
                 }
                 peaks = channels[0];
             }
+
+            if (this.hideChannel(channelIndex)) {
```
Comment (thijstriemstra) [inline src/drawer.multicanvas.js:499]: can you add a comment here explaining why it returns

File: src/wavesurfer.js [hunk 3]
```diff
@@ -1168,6 +1173,16 @@ export default class WaveSurfer extends util.Observer {
         this.drawBuffer();
     }
 
+    /**
+     * When splitChannels is true, hides channels from being drawn on the waveform.
+     *
```
Comment (thijstriemstra) [inline src/wavesurfer.js:1178]: i'm not able to visualize what happens here when reading that sentence. Can you rephrase/elaborate bit more? maybe also add an `@example` with small example as well to demonstrate it's use.

File: src/drawer.multicanvas.js (hunk 1 above)
Comment (win-chen) [reply src/drawer.multicanvas.js:395]: done

File: src/drawer.multicanvas.js (hunk 2 above)
Comment (win-chen) [reply src/drawer.multicanvas.js:500]: done. let me know if it's still unclear.

File: src/wavesurfer.js (hunk 3 above)
Comment (win-chen) [reply src/wavesurfer.js:1185]: Does this make better sense now?

**Comment by win-chen** [comment]
I've been trying to figure out why the appveyor build is failing but I can't seem to replicate the bug on local builds. Do you have any advice on how to debug this?

File: src/wavesurfer.js (hunk 3 above)
Comment (thijstriemstra) [reply src/wavesurfer.js:1185]: much better!

**Comment by thijstriemstra** [comment]
> I've been trying to figure out why the appveyor build is failing but I can't seem to replicate the bug on local builds. Do you have any advice on how to debug this?

Appveyor has a mind of it's own, don't worry about it.

File: src/wavesurfer.js [hunk 4]
```diff
@@ -1168,6 +1173,23 @@ export default class WaveSurfer extends util.Observer {
...
+     * Hide channels from being drawn on the waveform if splitting channels.
+     * 
+     * For example, if we want to draw only the peaks for the right stereo channel:
+     *
+     * const wavesurfer = new WaveSurfer.create({...splitChannels: true});
+     * wavesurfer.load('stereo_audio.mp3');
+     * 
+     * wavesurfer.setFilteredChannel([0]); <-- hide left channel peaks.
+     *
+     * @param {array} channelIndices Channels to be filtered out from drawing.
```
Comment (thijstriemstra) [inline src/wavesurfer.js:1186]: can you add an `@version 4.0.0` to this method?

Comment (win-chen) [reply src/wavesurfer.js:1186]: done and done.

**Comment by thijstriemstra** [comment]
> Really nice! I like this solution a lot better than the other PR. Thank you!

Do you want to drop the news in that pull request @katspaugh (and close it..)? Very unfortunate that contribution could not be used.

File: src/wavesurfer.js [hunk 5]
```diff
@@ -1168,6 +1173,24 @@ export default class WaveSurfer extends util.Observer {
         this.drawBuffer();
     }
 
+    /**
+     * Hide channels from being drawn on the waveform if splitting channels.
+     * 
+     * For example, if we want to draw only the peaks for the right stereo channel:
+     *
```
Comment (thijstriemstra) [inline src/wavesurfer.js:1180]: last nit: can you use an `@example` here, see https://jsdoc.app/tags-example.html

**Comment by MicBehrens** [comment]
Sorry for commenting on this old PR, but these new options are really awesome and i can't seem to get it working.
The following snippets works with splitting the channels, but it won't color the waves as stated in the channelColors options.
Can any of you tell me why?

```
        createAudioPlayer(recording) {
            this.audioPlayer = WaveSurfer.create({
                container: '#waveform',
                scrollParent: true,
                waveColor: '#D1DED3',
                progressColor: '#00AC13',
                barWidth: 1,
                barHeight: 10,
                barGap: 2,
                height: 70,
                hideScrollbar: true,
                plugins: [
                    WaveSurfer.cursor.create({
                        showTime: true,
                        opacity: 1,
                        customShowTimeStyle: {
                            'background-color': '#000000',
                            color: '#FFFFFF',
                            padding: '2px',
                            'font-size': '10px'
                        }
                    })
                ],
                splitChannels: true,
                splitChannelsOptions: {
                    overlay: false,
                    channelColors: {
                        0: {
                            progressColor: '#00AC13',
                            waveColor: '#D1DED3'
                        },
                        1: {
                            progressColor: '#3E81FF',
                            waveColor: '#D1DED3'
                        }
                    },
                    filterChannels: []
                }
            });
            this.audioPlayer.load(recording);
            this.audioPlayer.on('audioprocess', () => {
                this.current_time = this.audioPlayer.getCurrentTime();
            });
            this.audioPlayer.on('ready', () => {
                this.total_time = this.audioPlayer.getDuration();
                this.audioPlayer.play();
            });
            this.audioPlayer.on('error', (e) => {
                console.warn(e);
            });
        }
```
